[Bug]: Daily tests failing for python3.9-minimum on macosx

### What happened?

Passing (Aug 28): https://github.com/hdmf-dev/hdmf-zarr/actions/runs/17315377940/job/49157220442
Failing (Aug 29): https://github.com/hdmf-dev/hdmf-zarr/actions/runs/17339812047/job/49232135819

numcodecs 0.10.0 fails to build. There are no wheels available for arm64 on pypi.

The only differences I see are that the newer failing build uses a newer macos image which has a newer default version of xcode (16.4 from 16.0).

I can reproduce the failure on my computer (if I clear the built wheel for numcodecs 0.10.0 from my pip cache). I cannot reproduce the previous success. I cannot tell if numcodecs 0.10.0 was being built from tarball in the passing build, but it must be.

Options:
- Investigate the root cause further
- Bump up minimum numcodecs version to one that has a prebuilt wheel for macos arm64
- Bump up minimum python version to 3.10 and see if that helps

### Steps to Reproduce

```shell
See above
```

### Traceback

```shell
Run tox -e py39-minimum
  tox -e py39-minimum
  shell: /bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    pythonLocation: /Users/runner/hostedtoolcache/Python/3.9.13/arm64
    PKG_CONFIG_PATH: /Users/runner/hostedtoolcache/Python/3.9.13/arm64/lib/pkgconfig
    Python_ROOT_DIR: /Users/runner/hostedtoolcache/Python/3.9.13/arm64
    Python2_ROOT_DIR: /Users/runner/hostedtoolcache/Python/3.9.13/arm64
    Python3_ROOT_DIR: /Users/runner/hostedtoolcache/Python/3.9.13/arm64
py39-minimum: install_deps> python -m pip install -r requirements-min.txt
  error: subprocess-exited-with-error
  
  × Building wheel for numcodecs (pyproject.toml) did not run successfully.
  │ exit code: 1
  ╰─> [1151 lines of output]
      /private/var/folders/x7/ch5v91h56_zbvbd1y2f600dm0000gn/T/pip-build-env-m_48dpf9/overlay/lib/python3.9/site-packages/setuptools/dist.py:759: SetuptoolsDeprecationWarning: License classifiers are deprecated.
      !!
      
              ********************************************************************************
              Please consider removing the following classifiers in favor of a SPDX license expression:
      
              License :: OSI Approved :: MIT License
      
              See https://packaging.python.org/en/latest/guides/writing-pyproject-toml/#license for details.
              ********************************************************************************
      
      !!
        self._finalize_license_expression()
      [numcodecs] setting up Blosc extension
      [numcodecs] compiling Blosc extension without SSE2 support
      [numcodecs] compiling Blosc extension without AVX2 support
      [numcodecs] setting up Zstandard extension
      [numcodecs] setting up LZ4 extension
      [numcodecs] setting up compat extension
      [numcodecs] setting up shuffle extension
      [numcodecs] setting up vlen extension
      running bdist_wheel
      running build
      running build_py
      creating build/lib.macosx-10.9-universal2-cpython-39/numcodecs
      copying ./numcodecs/lzma.py -> build/lib.macosx-10.9-universal2-cpython-39/numcodecs
      copying ./numcodecs/astype.py -> build/lib.macosx-10.9-universal2-cpython-39/numcodecs
      copying ./numcodecs/gzip.py -> build/lib.macosx-10.9-universal2-cpython-39/numcodecs
      copying ./numcodecs/base64.py -> build/lib.macosx-10.9-universal2-cpython-39/numcodecs
      copying ./numcodecs/bz2.py -> build/lib.macosx-10.9-universal2-cpython-39/numcodecs
      copying ./numcodecs/version.py -> build/lib.macosx-10.9-universal2-cpython-39/numcodecs
      copying ./numcodecs/compat.py -> build/lib.macosx-10.9-universal2-cpython-39/numcodecs
      copying ./numcodecs/checksum32.py -> build/lib.macosx-10.9-universal2-cpython-39/numcodecs
      copying ./numcodecs/ndarray_like.py -> build/lib.macosx-10.9-universal2-cpython-39/numcodecs
  Downloading jsonschema-4.25.1-py3-none-any.whl.metadata (7.6 kB)
Collecting pandas>=1.0.5 (from hdmf==3.14.5->-r requirements-min.txt (line 3))
  Downloading pandas-2.3.2-cp39-cp39-macosx_11_0_arm64.whl.metadata (91 kB)
Collecting ruamel-yaml>=0.16 (from hdmf==3.14.5->-r requirements-min.txt (line 3))
  Downloading ruamel.yaml-0.18.15-py3-none-any.whl.metadata (25 kB)
Collecting scipy>=1.4 (from hdmf==3.14.5->-r requirements-min.txt (line 3))
  Downloading scipy-1.13.1-cp39-cp39-macosx_12_0_arm64.whl.metadata (60 kB)
Collecting asciitree (from zarr==2.18.0->-r requirements-min.txt (line 4))
  Downloading asciitree-0.3.3.tar.gz (4.0 kB)
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
Collecting fasteners (from zarr==2.18.0->-r requirements-min.txt (line 4))
  Downloading fasteners-0.20-py3-none-any.whl.metadata (4.8 kB)
Collecting typing-extensions>=3.7.4 (from numcodecs==0.10.0->-r requirements-min.txt (line 6))
  Using cached typing_extensions-4.15.0-py3-none-any.whl.metadata (3.3 kB)
Collecting python-dateutil>=2.7.3 (from pynwb==2.8.3->-r requirements-min.txt (line 7))
  Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl.metadata (8.4 kB)
Collecting attrs>=22.2.0 (from jsonschema>=2.6.0->hdmf==3.14.5->-r requirements-min.txt (line 3))
  Downloading attrs-25.3.0-py3-none-any.whl.metadata (10 kB)
Collecting jsonschema-specifications>=2023.03.6 (from jsonschema>=2.6.0->hdmf==3.14.5->-r requirements-min.txt (line 3))
  Downloading jsonschema_specifications-2025.4.1-py3-none-any.whl.metadata (2.9 kB)
Collecting referencing>=0.28.4 (from jsonschema>=2.6.0->hdmf==3.14.5->-r requirements-min.txt (line 3))
  Downloading referencing-0.36.2-py3-none-any.whl.metadata (2.8 kB)
Collecting rpds-py>=0.7.1 (from jsonschema>=2.6.0->hdmf==3.14.5->-r requirements-min.txt (line 3))
  Downloading rpds_py-0.27.1-cp39-cp39-macosx_11_0_arm64.whl.metadata (4.2 kB)
Collecting pytz>=2020.1 (from pandas>=1.0.5->hdmf==3.14.5->-r requirements-min.txt (line 3))
  Downloading pytz-2025.2-py2.py3-none-any.whl.metadata (22 kB)
Collecting tzdata>=2022.7 (from pandas>=1.0.5->hdmf==3.14.5->-r requirements-min.txt (line 3))
  Downloading tzdata-2025.2-py2.py3-none-any.whl.metadata (1.4 kB)
Collecting six>=1.5 (from python-dateutil>=2.7.3->pynwb==2.8.3->-r requirements-min.txt (line 7))
  Downloading six-1.17.0-py2.py3-none-any.whl.metadata (1.7 kB)
Collecting ruamel.yaml.clib>=0.2.7 (from ruamel-yaml>=0.16->hdmf==3.14.5->-r requirements-min.txt (line 3))
  Downloading ruamel.yaml.clib-0.2.12-cp39-cp39-macosx_12_0_arm64.whl.metadata (1.2 kB)
Downloading hdmf-3.14.5-py3-none-any.whl (338 kB)
Downloading zarr-2.18.0-py3-none-any.whl (209 kB)
Downloading numpy-1.24.0-cp39-cp39-macosx_11_0_arm64.whl (13.9 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 13.9/13.9 MB 110.2 MB/s  0:00:00
Downloading pynwb-2.8.3-py3-none-any.whl (1.4 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.4/1.4 MB 122.6 MB/s  0:00:00
Downloading threadpoolctl-3.1.0-py3-none-any.whl (14 kB)
Downloading h5py-3.14.0-cp39-cp39-macosx_11_0_arm64.whl (2.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.8/2.8 MB 20.3 MB/s  0:00:00
Downloading jsonschema-4.25.1-py3-none-any.whl (90 kB)
Downloading attrs-25.3.0-py3-none-any.whl (63 kB)
Downloading jsonschema_specifications-2025.4.1-py3-none-any.whl (18 kB)
Downloading pandas-2.3.2-cp39-cp39-macosx_11_0_arm64.whl (10.8 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 111.3 MB/s  0:00:00
Downloading python_dateutil-2.9.0.post0-py2.py3-none-any.whl (229 kB)
Downloading pytz-2025.2-py2.py3-none-any.whl (509 kB)
Downloading referencing-0.36.2-py3-none-any.whl (26 kB)
Downloading rpds_py-0.27.1-cp39-cp39-macosx_11_0_arm64.whl (354 kB)
Downloading ruamel.yaml-0.18.15-py3-none-any.whl (119 kB)
Downloading ruamel.yaml.clib-0.2.12-cp39-cp39-macosx_12_0_arm64.whl (131 kB)
Downloading scipy-1.13.1-cp39-cp39-macosx_12_0_arm64.whl (30.3 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 30.3/30.3 MB 88.2 MB/s  0:00:00
Downloading six-1.17.0-py2.py3-none-any.whl (11 kB)
Using cached typing_extensions-4.15.0-py3-none-any.whl (44 kB)
Downloading tzdata-2025.2-py2.py3-none-any.whl (347 kB)
Downloading fasteners-0.20-py3-none-any.whl (18 kB)
Building wheels for collected packages: numcodecs, asciitree
  Building wheel for numcodecs (pyproject.toml): started
  Building wheel for numcodecs (pyproject.toml): finished with status 'error'
  Building wheel for asciitree (pyproject.toml): started
  Building wheel for asciitree (pyproject.toml): finished with status 'done'
  Created wheel for asciitree: filename=asciitree-0.3.3-py3-none-any.whl size=5096 sha256=9df9580bac88ba664d96a9fc1bc32b0d79759b89bf02bb53020a3c376f31f97b
  Stored in directory: /Users/runner/Library/Caches/pip/wheels/07/00/bc/937e878be0d781a569792a1e0e5acf149e463d4536453ec978
Successfully built asciitree
Failed to build numcodecs
py39-minimum: exit 1 (17.82 seconds) /Users/runner/work/hdmf-zarr/hdmf-zarr> python -m pip install -r requirements-min.txt pid=1208
  py39-minimum: FAIL code 1 (19.03 seconds)
  evaluation failed :( (19.06 seconds)
Error: Process completed with exit code 1.
```

### Operating System

macOS

### Python Version

3.9

### Package Versions

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
 - Renamed `ZarrIO.file` property to `ZarrIO._file` to be consistent with PyNWB validator. @rly [#294](https://github.com/hdmf-dev/hdmf-zarr/pull/294)
 - When loading namespaces with `ZarrIO`, all namespaces are now passed to `namespace_catalog.load_namespaces` to handle ordering of loading and detection of version conflicts, to be consistent with `HDF5IO`. @rly [#294](https://github.com/hdmf-dev/hdmf-zarr/pull/294)
 - Renamed test module `tests/unit/utils.py` to `tests/unit/helpers/utils.py`. [#292](https://github.com/hdmf-dev/hdmf-zarr/pull/292)
+- Bumped minimum required version of `numcodecs` to 0.12.0 to address installation issues with Python 3.9 on MacOS. @rly [#299](https://github.com/hdmf-dev/hdmf-zarr/pull/299)
 
 ### Added
 - Added `ZarrIO.load_namespaces_io` method to load namespaces from a given Zarr group into a given `NamespaceCatalog`. @rly [#292](https://github.com/hdmf-dev/hdmf-zarr/pull/292)

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -33,7 +33,7 @@ dependencies = [
     "hdmf>=4.1.0",
     "zarr>=2.18.0, <3.0",  # pin below 3.0 until HDMF-zarr supports zarr 3.0
     "numpy>=1.24.0",
-    "numcodecs>=0.11.0, <0.16.0",  # numcodecs 0.16.0 is not compatible with zarr<3
+    "numcodecs>=0.12.0, <0.16.0",  # numcodecs 0.16.0 is not compatible with zarr<3
     "pynwb>=2.8.3",
     "threadpoolctl>=3.1.0",
 ]

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -33,7 +33,7 @@ dependencies = [
     "hdmf>=4.1.0",
     "zarr>=2.18.0, <3.0",  # pin below 3.0 until HDMF-zarr supports zarr 3.0
     "numpy>=1.24.0",
-    "numcodecs>=0.10.0, <0.16.0",  # numcodecs 0.16.0 is not compatible with zarr<3
+    "numcodecs>=0.11.0, <0.16.0",  # numcodecs 0.16.0 is not compatible with zarr<3
     "pynwb>=2.8.3",
     "threadpoolctl>=3.1.0",
 ]

--- a/requirements-min.txt
+++ b/requirements-min.txt
@@ -3,6 +3,6 @@
 hdmf==4.1.0
 zarr==2.18.0
 numpy==1.24.0
-numcodecs==0.10.0
+numcodecs==0.11.0
 pynwb==2.8.3
 threadpoolctl==3.1.0

--- a/requirements-min.txt
+++ b/requirements-min.txt
@@ -3,6 +3,6 @@
 hdmf==4.1.0
 zarr==2.18.0
 numpy==1.24.0
-numcodecs==0.11.0
+numcodecs==0.12.0
 pynwb==2.8.3
 threadpoolctl==3.1.0

